File: invictus_android/android/src/main/kotlin/dev/zeekr/invictus_android/internal/app/LocalePicker.kt
package dev.zeekr.invictus_android.internal.app

import android.content.Context
import android.os.LocaleList
import dev.zeekr.invictus_android.InvictusObject
import java.lang.reflect.Proxy
import java.util.Locale

class LocalePicker : InvictusObject {
    abstract class LocaleSelectionListener : InvictusObject() {
        companion object {
            internal val clazz: Class<*> get() = Class.forName("com.android.internal.app.LocalePicker\$LocaleSelectionListener")
        }

        final override val obj: Any

        init {
            this.obj = Proxy.newProxyInstance(clazz.classLoader, arrayOf(clazz)) { _, _, args ->
                val locale = args[0] as Locale
                this.onLocaleSelected(locale)
            }
        }

        abstract fun onLocaleSelected(locale: Locale)
    }

    class LocaleInfo : InvictusObject, Comparable<LocaleInfo> {
        companion object {
            internal val clazz: Class<*> get() = Class.forName("com.android.internal.app.LocalePicker\$LocaleInfo")
        }

        override val obj: Any

        internal constructor(obj: Any) {
            this.obj = obj
        }

        constructor(label: String, locale: Locale) {
            this.obj = clazz.getConstructor(String::class.java, Locale::class.java)
                .newInstance(label, locale)
        }

        val label: String get() = clazz.getMethod("getLabel").invoke(obj) as String
        val locale: Locale get() = clazz.getMethod("getLocale").invoke(obj) as Locale

        override fun compareTo(other: LocaleInfo): Int {
            return clazz.getMethod("compareTo", clazz).invoke(this.obj, other.obj) as Int
        }
    }

    companion object {
        internal val clazz: Class<*> get() = Class.forName("com.android.internal.app.LocalePicker")

        fun getSystemAssetLocales(): Array<String> {
            val locales = clazz.getMethod("getSystemAssetLocales").invoke(null) as Array<*>
            return locales.filterIsInstance<String>().toTypedArray()
        }

        fun getSupportedLocales(context: Context): Array<String> {
            val locales = clazz.getMethod("getSupportedLocales", Context::class.java)
                .invoke(null, context) as Array<*>
            return locales.filterIsInstance<String>().toTypedArray()
        }

        fun getAllAssetLocales(context: Context, isInDeveloperMode: Boolean): List<LocaleInfo> {
            val objs =
                clazz.getMethod("getAllAssetLocales", Context::class.java, Boolean::class.java)
                    .invoke(null, context, isInDeveloperMode) as List<*>
            return objs.filterNotNull().map { LocaleInfo(it) }
        }

        fun setLocaleSelectionListener(listener: LocaleSelectionListener) {
            clazz.getMethod("setLocaleSelectionListener", LocaleSelectionListener.clazz)
                .invoke(null, listener.obj)
        }

        /**
         * Requests the system to update the system locale. Note that the system looks halted
         * for a while during the Locale migration, so the caller need to take care of it.
         *
         * @see #updateLocales(LocaleList)
         */
        fun updateLocale(locale: Locale) {
            clazz.getMethod("updateLocale", Locale::class.java).invoke(null, locale)
        }

        /**
         * Requests the system to update the list of system locales.
         * Note that the system looks halted for a while during the Locale migration,
         * so the caller need to take care of it.
         */
        fun updateLocales(locales: LocaleList) {
            clazz.getMethod("updateLocales", LocaleList::class.java).invoke(null, locales)
        }

        /**
         * Get the locale list.
         *
         * @return The locale list.
         */
        fun getLocales(): LocaleList {
            return clazz.getMethod("getLocales").invoke(null) as LocaleList
        }
    }

    override val obj: Any

    internal constructor(obj: Any) {
        this.obj = obj
    }
}

File: invictus_api/lib/src/internal/app/locale_picker.dart
import 'package:invictus_api/src/os.dart';
import 'package:invictus_api/src/util.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

abstract interface class LocalePicker$LocaleSelectionListener {
  factory LocalePicker$LocaleSelectionListener({
    required void Function(Locale locale) onLocaleSelected,
  }) => LocalePickerChannel.instance.createLocaleSelectionListener(
    onLocaleSelected: onLocaleSelected,
  );
}

abstract interface class LocalePicker$LocaleInfo
    implements Comparable<LocalePicker$LocaleInfo> {
  factory LocalePicker$LocaleInfo(String label, Locale locale) =>
      LocalePickerChannel.instance.createLocaleInfo(label, locale);

  String get label;
  Locale get locale;
}

abstract interface class LocalePicker {
  // TODO
  static List<String> getSystemAssetLocales() =>
      LocalePickerChannel.instance.getSystemAssetLocales();

  // TODO
  static List<String> getSupportedLocales() =>
      LocalePickerChannel.instance.getSupportedLocales();

  // TODO
  static List<LocalePicker$LocaleInfo> getAllAssetLocales({
    bool isInDeveloperMode = false,
  }) => LocalePickerChannel.instance.getAllAssetLocales(
    isInDeveloperMode: isInDeveloperMode,
  );

  // TODO
  static void setLocaleSelectionListener(
    LocalePicker$LocaleSelectionListener listener,
  ) => LocalePickerChannel.instance.setLocaleSelectionListener(listener);

  /// Requests the system to update the system locale. Note that the system looks halted
  /// for a while during the Locale migration, so the caller need to take care of it.
  ///
  /// @see #updateLocales(LocaleList)
  static void updateLocale(Locale locale) =>
      LocalePickerChannel.instance.updateLocale(locale);

  /// Requests the system to update the list of system locales.
  /// Note that the system looks halted for a while during the Locale migration,
  /// so the caller need to take care of it.
  static void updateLocales(LocaleList locales) =>
      LocalePickerChannel.instance.updateLocales(locales);

  /// Get the locale list.
  ///
  /// @return The locale list.
  static LocaleList getLocales() => LocalePickerChannel.instance.getLocales();
}

abstract base class LocalePickerChannel extends PlatformInterface {
  /// Constructs a [LocalePickerChannel].
  LocalePickerChannel() : super(token: _token);

  static final Object _token = Object();

  static LocalePickerChannel? _instance;

  /// The default instance of [LocalePickerChannel] to use.
  static LocalePickerChannel get instance =>
      ArgumentError.checkNotNull(_instance, 'instance');

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [LocalePickerChannel] when
  /// they register themselves.
  static set instance(LocalePickerChannel instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  LocalePicker$LocaleSelectionListener createLocaleSelectionListener({
    required void Function(Locale locale) onLocaleSelected,
  });
  LocalePicker$LocaleInfo createLocaleInfo(String label, Locale locale);

  List<String> getSystemAssetLocales();
  List<String> getSupportedLocales();
  List<LocalePicker$LocaleInfo> getAllAssetLocales({
    bool isInDeveloperMode = false,
  });
  void setLocaleSelectionListener(
    LocalePicker$LocaleSelectionListener listener,
  );
  void updateLocale(Locale locale);
  void updateLocales(LocaleList locales);
  LocaleList getLocales();
}
